src/compartment.c

```diff
@@ -139,192 +142,302 @@ comp_from_elf(char* filename, struct ConfigEntryPoint* entry_points, size_t entr
+
+        // Find additional library dependencies
+        Elf64_Dyn* comp_dyn_entries = malloc(comp_dynamic_shdr.sh_size);
+        do_pread((int) new_comp->fd, comp_dyn_entries,
+                 comp_dynamic_shdr.sh_size, comp_dynamic_shdr.sh_offset);
+
+        for (size_t i = 0;
+             i < comp_dynamic_shdr.sh_size / sizeof(Elf64_Dyn);
+             ++i)
+        {
+            if (comp_dyn_entries[i].d_tag == DT_NEEDED)
+            {
+                struct LibDependency* new_lib_dep =
+                    malloc(sizeof(struct LibDependency));
+                new_lib_dep->lib_name =
+                    malloc(strlen(&dyn_str_tbl[comp_dyn_entries[i].d_un.d_val]) + 1);
+                strcpy(
+                    new_lib_dep->lib_name,
+                    &dyn_str_tbl[comp_dyn_entries[i].d_un.d_val]);
+                new_comp->lib_deps_count += 1;
+                new_comp->lib_deps =
+                    realloc(new_comp->lib_deps,
+                            new_comp->lib_deps_count *
+                                sizeof(struct LibDependency));
+                new_comp->lib_deps[new_comp->lib_deps_count - 1] = new_lib_dep;
             }
+        }
+
+        free(dyn_str_tbl);
+        free(comp_dyn_entries);
+    }
 
-            new_comp->relas_cnt = comp_symtb_hdr.sh_size / sizeof(Elf64_Rela);
-            new_comp->relas = (uintptr_t*) malloc(new_comp->relas_cnt * sizeof(uintptr_t));
-            Elf64_Rela* comp_relas = (Elf64_Rela*) malloc(comp_symtb_hdr.sh_size);
-            pread_res = pread((int) new_comp->fd, comp_relas, comp_symtb_hdr.sh_size, comp_symtb_hdr.sh_offset);
-            assert(pread_res != -1);
+    // Find library dependencies in `COMP_LIBRARY_PATH`
+    for (size_t i = 0; i < new_comp->lib_deps_count; ++i)
+    {
+        struct LibDependency* curr_dep = new_comp->lib_deps[i];
+        // TODO move env var name to constant
+        printf("-- COMP_LIB_PATH %s\n", getenv("COMP_LIBRARY_PATH"));
+        assert(getenv("COMP_LIBRARY_PATH"));
+        char* lib_path =
+            find_in_dir(curr_dep->lib_name, getenv("COMP_LIBRARY_PATH"));
+        if (!lib_path)
+        {
+            printf("Could not find file for dependency %s!\n", curr_dep->lib_name);
+            exit(1);
+        }
+        curr_dep->lib_path = malloc(strlen(lib_path));
+        strcpy(curr_dep->lib_path, lib_path);
+        init_lib_dep_info(curr_dep, new_comp);
+        new_comp->mem_top =
+            curr_dep->lib_mem_base +
+            curr_dep->lib_segs[curr_dep->lib_segs_count - 1]->mem_top;
+    }
 
-            for (size_t j = 0; j < new_comp->relas_cnt; ++j)
+    // Find functions of interest, particularly entry points, and functions to
+    // intercept
+    Elf64_Shdr comp_strtb_hdr;
+    do_pread((int) new_comp->fd, &comp_strtb_hdr, sizeof(Elf64_Shdr),
+             comp_ehdr.e_shoff + comp_symtb_shdr.sh_link * sizeof(Elf64_Shdr));
+
+    // XXX Small trick here - we define the string table as a pointer
```

Comment:
Not sure I full understand this. Since the "string" type in C is `char *` a string (and any position in it) is a pointer to a char?

Reply:
Doesn't `pread` just read in bytes? I'm not sure there's anything clever going on here.

Reply:
Can I double check, when you mean "this" in the first message, do you mean line 320 or 323?

Reply:
I think this is the only remaining comment now.

Reply:
Heh, Github hid it from me.

I think what I'm saying here is I tried defining the string table as a `char**`, but I kept messing up the indexing. So it's easier for me to define it as a `char*`, since it more naturally maps to how symbol names are indexed via the symbol table.

I can try to rephrase this comment, if the above makes sense?